### PR TITLE
Patch maps better

### PR DESCRIPTION
**Pull request by treeowl**
Use the `containers` merge interface to improve efficiency.

Make strictness consistent.

### CODE REVIEW

**Comment by treeowl** [comment]
I did a fancy useless-deletion detection thing for `Map`; I can do the same for `IntMap` if that's how we want to go. It would also be possible to *try* to detect useless updates, but it'll be pretty unreliable for many important cases. We could add a `PatchMap` type that insists on its values being `Eq`, but would that be useful?

File: src/Reflex/Patch/Map.hs [hunk 1]
```diff
@@ -22,12 +25,34 @@ newtype PatchMap k v = PatchMap { unPatchMap :: Map k (Maybe v) }
...
+          (zipWithMaybeAMatched (\_k mv v -> Changed $! mv <|> Just v)) p old
+
+changedToMaybe :: Changed a -> Maybe a
+changedToMaybe (Unchanged _) = Nothing
+changedToMaybe (Changed a) = Just a
+
+data Changed a
+  = Unchanged a
+  | Changed a
+  deriving (Functor)
```
Comment (3noch) [inline src/Reflex/Patch/Map.hs]: Are there any speed benefits to making this `newtype Changed a = Changed (ChangedTag, a)`? Or does that, as I imagine, compile down to the same thing?

Comment (treeowl) [inline src/Reflex/Patch/Map.hs]: Should I change this to `data Changed a = Changed !Bool a` or similar? I'll have to look at Core to see if GHC will unbox it nicely in that case. Might be an improvement.

Comment (treeowl) [reply src/Reflex/Patch/Map.hs]: See my comment from a minute ago.

**Comment by cgibbard** [comment]
Haha, this kind of thing is exactly why I got you to make Data.Map.Merge in the first place, it's kind of funny that you'd also end up helping us replace our old awkward merges with it. :D

File: src/Reflex/Patch/Map.hs (hunk 1 above)
Comment (lpsmith) [reply src/Reflex/Patch/Map.hs]: Lazy pairs are kind of challenging to compile,  so yeah,  it's really between what @treeowl suggested and what's already written.  (But I don't think this issue should block anything on this MR.)

**Comment by Ericson2314** [comment]
Moved to https://github.com/reflex-frp/patch/pull/3

**Pull request closed without merging**